File: iosApp/iosApp/Core/Journey/JourneyComparison.swift
import Foundation

/// Plan companion helpers (foldables master plan, Plan contract): compare the
/// route alternatives on the differences that are actually there, and keep the
/// selected itinerary by identity rather than by list position.
///
/// Swift twin of Kotlin `com.syrmos.core.domain.journey.JourneyComparison`;
/// the fixtures in `JourneyComparisonTests` and `JourneyComparisonTest` must agree.
enum JourneyComparison {
    /// What one alternative has over the others. A lone alternative is neutral:
    /// nothing to compare against, so a single route is never decorated as fastest.
    struct Facts: Equatable {
        /// Shortest known duration, and at least one other option is slower.
        var fastest: Bool = false
        /// Fewest changes, and at least one other option has more.
        var fewestChanges: Bool = false
        /// Whole minutes behind the fastest known option; nil when this is the
        /// fastest, ties it, or its duration is unknown.
        var minutesSlowerThanFastest: Int? = nil
        /// Changes beyond the option with the fewest (0 when it is one).
        var extraChanges: Int = 0
    }

    /// Facts for every alternative, index-aligned with the inputs. `durations`
    /// are seconds (nil when untimed), `changes` the transfer counts.
    static func facts(durations: [Int?], changes: [Int]) -> [Facts] {
        precondition(durations.count == changes.count, "durations and changes must align")
        let n = durations.count
        if n < 2 { return Array(repeating: Facts(), count: n) }

        let known = durations.compactMap { $0 }
        let minDur = known.min()
        let maxDur = known.max()
        let durationsDiffer: Bool = {
            guard let lo = minDur, let hi = maxDur else { return false }
            return hi > lo
        }()
        let minCh = changes.min() ?? 0
        let changesDiffer = (changes.max() ?? 0) > minCh

        return durations.indices.map { i in
            let d = durations[i]
            var slower: Int? = nil
            if let d, let lo = minDur, d > lo {
                let m = (d - lo + 30) / 60
                slower = m > 0 ? m : nil
            }
            return Facts(
                fastest: durationsDiffer && d != nil && d == minDur,
                fewestChanges: changesDiffer && changes[i] == minCh,
                minutesSlowerThanFastest: slower,
                extraChanges: changes[i] - minCh
            )
        }
    }
}

/// Selected-itinerary identity across a results refresh (twin of Kotlin
/// `JourneySelection`). Keyed by option id, never by array index.
enum JourneySelection {
    /// The id to keep selected: `previous` when still offered, else the first id, else nil.
    static func retain(previous: String?, ids: [String]) -> String? {
        if let previous, ids.contains(previous) { return previous }
        return ids.first
    }

    /// Position of `selected` in `ids`, or nil when it is not offered.
    static func index(of selected: String?, in ids: [String]) -> Int? {
        guard let selected else { return nil }
        return ids.firstIndex(of: selected)
    }
}

File: core/domain/src/commonMain/kotlin/com/syrmos/core/domain/journey/JourneyComparison.kt
package com.syrmos.core.domain.journey

/**
 * Plan companion helpers (foldables master plan, Plan contract): compare the
 * route alternatives on the differences that are actually there, and keep the
 * selected itinerary by identity rather than by list position.
 *
 * Both are pure and language-neutral. iOS carries a twin in
 * `iosApp/Core/Journey/JourneyComparison.swift`; the fixtures in
 * `JourneyComparisonTest` and `JourneyComparisonTests` must agree.
 */
object JourneyComparison {
    /**
     * What one alternative has over the others.
     *
     * - [fastest]: shortest known duration, and at least one other option is slower.
     * - [fewestChanges]: fewest changes, and at least one other option has more.
     * - [minutesSlowerThanFastest]: whole minutes behind the fastest known
     *   option, null when this is the fastest, ties it, or its duration is unknown.
     * - [extraChanges]: changes beyond the option with the fewest (0 when it is one).
     *
     * A lone alternative gets neutral facts: there is nothing to compare against,
     * so a single route is never decorated as "fastest".
     */
    data class Facts(
        val fastest: Boolean = false,
        val fewestChanges: Boolean = false,
        val minutesSlowerThanFastest: Int? = null,
        val extraChanges: Int = 0,
    )

    /**
     * Facts for every alternative, index-aligned with the inputs. [durations] are
     * seconds (null when the option has no timed duration), [changes] the transfer
     * counts. The two lists must be the same length.
     */
    fun facts(durations: List<Int?>, changes: List<Int>): List<Facts> {
        require(durations.size == changes.size) { "durations and changes must align" }
        val n = durations.size
        if (n < 2) return List(n) { Facts() }

        val known = durations.filterNotNull()
        val minDur = known.minOrNull()
        val maxDur = known.maxOrNull()
        val durationsDiffer = minDur != null && maxDur != null && maxDur > minDur
        val minCh = changes.min()
        val changesDiffer = changes.max() > minCh

        return durations.indices.map { i ->
            val d = durations[i]
            val slower = if (d != null && minDur != null && d > minDur) {
                ((d - minDur + 30) / 60).takeIf { it > 0 }
            } else null
            Facts(
                fastest = durationsDiffer && d != null && d == minDur,
                fewestChanges = changesDiffer && changes[i] == minCh,
                minutesSlowerThanFastest = slower,
                extraChanges = changes[i] - minCh,
            )
        }
    }
}

/**
 * Selected-itinerary identity across a results refresh. Alternatives are keyed
 * by the option id (legs plus schedule identity), so a re-plan that reorders or
 * replaces options keeps the traveller's choice when it still exists and falls
 * back to the first option otherwise. Never an array index.
 */
object JourneySelection {
    /** The id to keep selected: [previous] when still offered, else the first id, else null. */
    fun retain(previous: String?, ids: List<String>): String? =
        if (previous != null && previous in ids) previous else ids.firstOrNull()

    /** Position of [selected] in [ids], or null when it is not offered. */
    fun index(selected: String?, ids: List<String>): Int? =
        selected?.let { s -> ids.indexOf(s).takeIf { it >= 0 } }
}
